ui/framework/src/test/imodel-components/VisibilityTreeEventHandler.test.ts

```diff
@@ -0,0 +1,119 @@
+  const dispose = sinon.fake();
+  const onVisibilityChange = new BeEvent<VisibilityChangeListener>();
+
+  const visibilityHandler: IVisibilityHandler = {
+    changeVisibility,
+    getVisibilityStatus,
+    onVisibilityChange,
+    dispose,
+  };
+
+  const testVisibilityStatus: VisibilityStatus = {
+    state: "visible",
+    isDisabled: false,
+  };
+
+  beforeEach(async () => {
+    modelSourceMock.reset();
+    modelMock.reset();
+    nodeLoaderMock.reset();
+    dataProviderMock.reset();
+    selectionHandlerMock.reset();
+    changeVisibility.resetHistory();
+    getVisibilityStatus.reset();
+    dispose.resetHistory();
+
+    getVisibilityStatus.returns(testVisibilityStatus);
+    modelMock.setup((x) => x.getNode(moq.It.isAny())).returns(() => createSimpleTreeModelNode());
+    modelSourceMock.setup((x) => x.onModelChanged).returns(() => new BeUiEvent<[TreeModel, TreeModelChanges]>());
+    modelSourceMock.setup((x) => x.getModel()).returns(() => modelMock.object);
+    nodeLoaderMock.setup((x) => x.dataProvider).returns(() => dataProviderMock.object);
+    nodeLoaderMock.setup((x) => x.modelSource).returns(() => modelSourceMock.object);
+  });
+
+  const createHandler = (partialProps?: Partial<VisibilityTreeEventHandlerParams>): VisibilityTreeEventHandler => {
+    if (!partialProps)
+      partialProps = {};
+    const props: VisibilityTreeEventHandlerParams = {
+      visibilityHandler: partialProps.visibilityHandler || undefined,
+      nodeLoader: partialProps.nodeLoader || nodeLoaderMock.object,
+      selectionHandler: partialProps.selectionHandler || selectionHandlerMock.object,
+    };
+    return new VisibilityTreeEventHandler(props);
+  };
+
+  describe("onChangeVisibility", () => {
+    it("calls getVisibilityStatus() only on the nodes, whose visibility status is not known when onChangeVisibility event is raised with only visibilityStatus", async () => {
+      const visibilityStatus: Map<string, VisibilityStatus> = new Map([
+        ["testId2", testVisibilityStatus],
+      ]);
+
+      const treeModelNodes = [
+        createSimpleTreeModelNode("testId1"),
+        createSimpleTreeModelNode("testId2"),
+        createSimpleTreeModelNode("testId3"),
+      ];
+
+      modelMock.setup((x) => x.iterateTreeModelNodes()).returns(() => treeModelNodes[Symbol.iterator]());
+
+      await using(createHandler({ visibilityHandler }), async (_) => {
+        onVisibilityChange.raiseEvent(undefined, visibilityStatus);
+      });
+      await TestUtils.flushAsyncOperations();
+
+      // getVisibilityStatus() is called 3 times from VisibilityTreeEventHandler constructor and 2 times by raising onVisibilityChange event.
+      expect(getVisibilityStatus.callCount).to.eq(5);
+    });
+
+    it("calls getVisibilityStatus() when onChangeVisibility event is raised with only nodeIds", async () => {
+      await using(createHandler({ visibilityHandler }), async (_) => {
+        onVisibilityChange.raiseEvent(["testId1", "testId2"]);
```

Comment:
Shouldn't there be `flushAsyncOperations` after event is raised?

Reply:
Fixed it